File: algorithms/08_r/dp/knapsack.R
knapsack <- function(weights, values, capacity) {
  n <- length(weights)
  dp <- rep(0, capacity + 1)
  
  for (i in 1:n) {
    if (weights[i] <= capacity) {
      for (w in capacity:weights[i]) {
        dp[w + 1] <- max(dp[w + 1], dp[w - weights[i] + 1] + values[i])
      }
    }
  }
  return(dp[capacity + 1])
}

File: algorithms/05_csharp/dp/knapsack.cs
using System;

namespace Algorithms.DynamicProgramming
{
    public static class Knapsack
    {
        public static int Solve(int[] weights, int[] values, int capacity)
        {
            int n = weights.Length;
            int[] dp = new int[capacity + 1];

            for (int i = 0; i < n; i++)
            {
                for (int w = capacity; w >= weights[i]; w--)
                {
                    dp[w] = Math.Max(dp[w], dp[w - weights[i]] + values[i]);
                }
            }

            return dp[capacity];
        }
    }
}
